### ISSUE TITLE
ttnn.reshape (block-sharded, tiled): stale scratch buffer leaks input data into output tile-padding rows

### ISSUE THREAD

**Issue by rpavlovicTT** [tenstorrent/tt-metal#42840]
## Summary

`ttnn.reshape` with block-sharded tiled tensors corrupts the output's tile-padding rows with real input data. The padding rows (e.g. seq rows 197–223 per batch after reshaping to `[8, 197, 2304]`) must be zero but contain values from earlier input tiles.

## Repro

Shape from ViT-base QKV projection:
- Input: `[1576, 2304]` block-sharded, shard `[224, 288]` on 8×8
- Output: `[8, 197, 2304]` block-sharded, same shard

The test below fills each input tile-row `t` with the sentinel value `t+1` (integers 1–50), reshapes, converts to interleaved, reads back the padded storage `[8, 224, 2304]`, and checks that rows 197–223 per batch are zero:

```python
import torch, ttnn

BS = ttnn.MemoryConfig(
    ttnn.TensorMemoryLayout.BLOCK_SHARDED, ttnn.BufferType.L1,
    ttnn.ShardSpec(
        ttnn.CoreRangeSet([ttnn.CoreRange(ttnn.CoreCoord(0, 0), ttnn.CoreCoord(7, 7))]),
        [224, 288], ttnn.ShardOrientation.ROW_MAJOR))

def test_reshape_block_sharded_does_not_leak_into_padding(device):
    PROBE = {float(t + 1) for t in range((1576 + 31) // 32)}  # {1.0 … 50.0}

    x = torch.zeros(1576, 2304, dtype=torch.bfloat16)
    for t in range(50):
        x[t * 32 : min((t + 1) * 32, 1576), :] = float(t + 1)

    tt = ttnn.from_torch(x, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT,
                         device=device, memory_config=BS)
    tt_r = ttnn.reshape(tt, [8, 197, 2304], memory_config=BS)
    tt_il = ttnn.to_memory_config(
        tt_r, ttnn.MemoryConfig(ttnn.TensorMemoryLayout.INTERLEAVED, ttnn.BufferType.L1))

    padded = tt_il.to_torch_with_padded_shape().to(torch.float32)
    padding = padded[:, 197:, :]

    leaks = {}
    for b in range(8):
        hits = [v for v in torch.unique(padding[b][padding[b] != 0]).tolist() if v in PROBE]
        if hits:
            leaks[b] = [int(v) - 1 for v in hits]
    assert not leaks, f"Leaked input tile-rows into output padding: {leaks}"
```

**Observed failure:**
```
output batch 0: contains data from input tile-row(s) [5]  (input rows 160-191)
output batch 1: contains data from input tile-row(s) [11, 12] (input rows 352-415)
...
output batch 7: contains data from input tile-row(s) [48, 49] (input rows 1536-1575)
```

## Root cause

In `writer_reshape_tiled.cpp` the scratch tile buffer (`cb_id_working`) is reserved once and reused across all output tiles without being cleared between iterations.

For **full** output tiles the reshape map covers all 1024 positions, so prior content is always overwritten — no problem.

For **partial** output tiles (where the logical dimension isn't tile-aligned — e.g. `197 % 32 = 5`, so the last tile-row per batch has only 5 logical rows), the map only covers positions 0–4 within the tile. Positions 5–31 (output rows 197–223) are never written, so they retain data from the previous full tile. The writer then DMA-writes the entire 32×32 scratch to the output buffer, propagating the stale data.

## Proposed fix

Two-pass scan of the map entries in the writer kernel:

1. **Pass 1** (pure L1 reads): sum `num_elements` across all map entries to compute `covered`.
2. If `covered < tile_elements`: call `fill_with_val` to zero the scratch before copying.
3. **Pass 2**: existing segment copy logic, unchanged.

Full tiles (`covered == tile_elements`) skip the zero entirely. Partial tiles pay a `fill_with_val` only when needed.

**File:** `ttnn/cpp/ttnn/operations/data_movement/reshape_view/device/device/dataflow/writer_reshape_tiled.cpp`
